### PR TITLE
Fix protocol parsing

### PR DESCRIPTION
**Pull request by dpgraham**
* Some problems with parsing `{protocol: <protocol>, value: <value>}`
  * `(!_.isNil(driverRes.value)` was incorrect check because `value` can be undefined or null; replaced it with `_.has(driverRes, 'value')` to check that the `value` property was set
  *  ```
       if (driverRes.value) {
          driverRes = driverRes.value;
        } 
      ``` 

      was also incorrect because `driverRes.value` may be false-y

* Moved logic into it's own function and then added tests to cover the problem cases

### CODE REVIEW

File: lib/protocol/protocol.js [hunk 1]
```diff
@@ -451,5 +448,21 @@ async function doJwpProxy (driver, req, res) {
   }
 }
 
+/**
+ * Check a driver command respond and see if the protocol and value was passed in
+ * @param {Object} driverRes
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: missing `return` description and `driverRes`

File: lib/protocol/protocol.js [hunk 2]
```diff
@@ -451,5 +448,21 @@ async function doJwpProxy (driver, req, res) {
...
+ */
+function parseProtocol (driverRes) {
+  if (_.isObject(driverRes) && _.isString(driverRes.protocol)) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: isPlainObject?

Comment (mykola-mokhnach) [reply lib/protocol/protocol.js]: I'd rather replace it with 

```
if (!_.isPlainObject(driverRes) || !_.isString(driverRes.protocol)) {
  return null;
}
```

File: lib/protocol/protocol.js (hunk 1 above)
Comment (mykola-mokhnach) [reply lib/protocol/protocol.js]: this function should also have `@throws` description

File: lib/protocol/protocol.js [hunk 3]
```diff
@@ -451,5 +448,21 @@ async function doJwpProxy (driver, req, res) {
...
+    if (_.has(driverRes, 'value')) {
+      let isW3C = driverRes.protocol === BaseDriver.DRIVER_PROTOCOL.W3C;
+      let isMJSONWP = driverRes.protocol === BaseDriver.DRIVER_PROTOCOL.MJSONWP;
+      return {isW3C, isMJSONWP, value: driverRes.value};
+    } else if (_.isError(driverRes.error)) {
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: else if redundant

**Comment by dpgraham** [comment]
Fixes pushed

File: lib/protocol/protocol.js [hunk 4]
```diff
@@ -451,5 +448,26 @@ async function doJwpProxy (driver, req, res) {
   }
 }
 
+/**
+ * Check a driver command respond and see if the protocol and value was passed in
+ * @param {Object} driverRes Response returned by `executeCommand` in an inner driver
+ * @throws Will throw an error if driverRes.error is provided
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: how about returns?

File: lib/protocol/protocol.js [hunk 5]
```diff
@@ -451,5 +448,26 @@ async function doJwpProxy (driver, req, res) {
...
+
+  if (_.has(driverRes, 'value')) {
+    let isW3C = driverRes.protocol === BaseDriver.DRIVER_PROTOCOL.W3C;
+    let isMJSONWP = driverRes.protocol === BaseDriver.DRIVER_PROTOCOL.MJSONWP;
+    return {isW3C, isMJSONWP, value: driverRes.value};
+  }
+
+  if (_.isError(driverRes.error)) {
+    throw driverRes.error;
+  }
```
Comment (mykola-mokhnach) [inline lib/protocol/protocol.js]: I prefer to explicitly return null, since it improves readability

**Comment by dpgraham** [comment]
Was published.

**Pull request merged**